File: Assets/Scripts/Core/BudgetSystem.cs
using System;
using System.Collections.Generic;

namespace MOVBattle.Core
{
    public sealed class BudgetSystem
    {
        private readonly Dictionary<TeamId, int> _remainingByTeam;
        private readonly int _maxUnitsPerTeam;

        public BudgetSystem(int teamABudget, int teamBBudget, int maxUnitsPerTeam)
        {
            _remainingByTeam = new Dictionary<TeamId, int>
            {
                [TeamId.TeamA] = Math.Max(0, teamABudget),
                [TeamId.TeamB] = Math.Max(0, teamBBudget)
            };

            _maxUnitsPerTeam = Math.Max(1, maxUnitsPerTeam);
        }

        public int GetRemaining(TeamId team)
        {
            return _remainingByTeam[team];
        }

        public bool CanAfford(TeamId team, int cost)
        {
            if (cost < 0)
            {
                return false;
            }

            return _remainingByTeam[team] >= cost;
        }

        public bool CanAddUnit(TeamId team, int existingCount)
        {
            if (existingCount < 0)
            {
                return false;
            }

            return existingCount < _maxUnitsPerTeam;
        }

        public bool TrySpend(TeamId team, int cost)
        {
            if (!CanAfford(team, cost))
            {
                return false;
            }

            _remainingByTeam[team] -= cost;
            return true;
        }

        public void Refund(TeamId team, int amount)
        {
            if (amount <= 0)
            {
                return;
            }

            _remainingByTeam[team] += amount;
        }
    }
}

File: web/src/core/budgetSystem.ts
export class BudgetSystem {
  private _remaining: [number, number];
  private _maxUnitsPerTeam: number;

  constructor(teamABudget: number, teamBBudget: number, maxUnitsPerTeam: number) {
    this._remaining = [Math.max(0, teamABudget), Math.max(0, teamBBudget)];
    this._maxUnitsPerTeam = Math.max(1, maxUnitsPerTeam);
  }

  getRemaining(team: number): number {
    return this._remaining[team];
  }

  canAfford(team: number, cost: number): boolean {
    return cost >= 0 && this._remaining[team] >= cost;
  }

  canAddUnit(team: number, existingCount: number): boolean {
    return existingCount >= 0 && existingCount < this._maxUnitsPerTeam;
  }

  trySpend(team: number, cost: number): boolean {
    if (!this.canAfford(team, cost)) return false;
    this._remaining[team] -= cost;
    return true;
  }

  refund(team: number, amount: number): void {
    if (amount <= 0) return;
    this._remaining[team] += amount;
  }
}
